File: tools/src/import/rosterizer.ts
/**
 * Rosterizer adapter: lower a Rosterizer roster JSON payload to a
 * {@link ParsedRoster}.
 *
 * Rosterizer (https://rosterizer.com) stores a roster as a `Roster` envelope
 * with a recursive `Asset` tree under `snapshot` (or `history.present.roster`
 * as a fallback). Every entity — faction, detachment, unit, weapon, ability,
 * enhancement — is an `Asset` keyed by `Classification§Designation` (e.g.
 * `"Unit§Tactical Squad"`). Children sit under `assets.included` (game pieces)
 * and `assets.traits` (modifiers, abilities, markers).
 *
 * The schema is rulebook-agnostic, so the actual `Classification` strings come
 * from whichever Rosterizer rulebook authored the roster. The constants below
 * encode the 40K convention used by the 40kdc reference rulebook; tune
 * them here without touching parser logic if a real export disagrees.
 *
 * **IP safety**: the walk reads an ALLOWLIST — `item`, `designation`, `name`,
 * `classification`, `quantity`, `meta.points`, `stats.Points.value`,
 * `aspects.Visibility`, and the recursive `assets.included`/`assets.traits`
 * children. Prose-bearing fields — `text`, `description`, `rules`, ability
 * `stats`, `_layers`, `lineage`, `processed`, `classIdentity`, `bareResourceKey`
 * — are never touched, so the importer's output is free of copyrighted prose
 * by construction.
 *
 * @packageDocumentation
 */
import type { FormatAdapter } from "./adapter.js";
import type { ParsedRoster, ParsedUnit, ParsedWargear } from "./types.js";

// --- 40K rulebook Classification§Designation conventions. -------------------
// These pin the strings the adapter looks for. Tune them in one place if a
// real Rosterizer export uses different labels.

const CLS_ROSTER = "Roster";
const CLS_FACTION = "Faction";
const CLS_DETACHMENT = "Detachment";
const CLS_UNIT = "Unit";
const CLS_SQUAD = "Squad"; // alternative unit class some rulebooks use
const CLS_WEAPON = "Weapon";
const CLS_ENHANCEMENT = "Enhancement";
const CLS_BATTLE_SIZE = "Battle Size";
const CLS_TRAIT = "Trait";
const DSG_WARLORD = "Warlord";
const CHAR_CLASSIFICATIONS = new Set(["Character", "Epic Hero"]);

const POINTS_STAT_KEYS = ["Points", "Pts"];
const POINTS_LIMIT = /(\d[\d,]*)\s*Point/i;

// --- Structural views ------------------------------------------------------

interface RawStat {
  value?: unknown;
}
interface RawAspects {
  Visibility?: unknown;
}
interface RawAssetChildren {
  included?: unknown;
  traits?: unknown;
}
interface RawAsset {
  item?: unknown;
  name?: unknown;
  designation?: unknown;
  classification?: unknown;
  quantity?: unknown;
  aspects?: unknown;
  assets?: unknown;
  meta?: unknown;
  stats?: unknown;
  keywords?: unknown;
}
interface RawRulebook {
  name?: unknown;
  game?: unknown;
  publisher?: unknown;
  url?: unknown;
}
interface RawHistoryItem {
  roster?: unknown;
  note?: unknown;
}
interface RawHistory {
  present?: unknown;
}
interface RawEnvelope {
  rulebook?: unknown;
  snapshot?: unknown;
  history?: unknown;
  slug?: unknown;
  authors?: unknown;
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function asObject(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function asString(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function asNumber(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string") {
    const n = Number.parseFloat(value);
    return Number.isFinite(n) ? n : null;
  }
  return null;
}

/** Split `Classification§Designation` into its two halves. Falls back to the
 * raw `classification`/`designation` fields when `item` is absent. */
function splitItem(asset: RawAsset): { classification: string; designation: string } {
  const item = asString(asset.item);
  if (item !== null) {
    const idx = item.indexOf("§"); // §
    if (idx >= 0) {
      return {
        classification: item.slice(0, idx),
        designation: item.slice(idx + 1),
      };
    }
  }
  return {
    classification: asString(asset.classification) ?? "",
    designation: asString(asset.designation) ?? "",
  };
}

/** A user-facing display name for an asset: `name` override beats the
 * designation parsed out of the `item` key. */
function displayName(asset: RawAsset): string {
  return asString(asset.name) ?? splitItem(asset).designation;
}

function quantity(asset: RawAsset): number {
  const n = asNumber(asset.quantity);
  return n !== null && n > 0 ? Math.trunc(n) : 1;
}

function included(asset: RawAsset): RawAsset[] {
  const a = asset.assets as RawAssetChildren | undefined;
  return asArray(a?.included) as RawAsset[];
}

function traits(asset: RawAsset): RawAsset[] {
  const a = asset.assets as RawAssetChildren | undefined;
  return asArray(a?.traits) as RawAsset[];
}

/** Points cost from `stats.Points.value` (or aliases) / `meta.points`, or null. */
function pointsOf(asset: RawAsset): number | null {
  const stats = asObject(asset.stats);
  if (stats) {
    for (const key of POINTS_STAT_KEYS) {
      const stat = asObject(stats[key]) as RawStat | null;
      if (stat) {
        const v = asNumber(stat.value);
        if (v !== null) return Math.trunc(v);
      }
    }
  }
  const meta = asObject(asset.meta);
  if (meta) {
    const v = asNumber(meta.points);
    if (v !== null) return Math.trunc(v);
  }
  return null;
}

/** Depth-first visit of an asset and every included/trait descendant. */
function walk(asset: RawAsset, visit: (a: RawAsset) => void): void {
  visit(asset);
  for (const child of included(asset)) walk(child, visit);
  for (const child of traits(asset)) walk(child, visit);
}

function classOf(asset: RawAsset): string {
  return splitItem(asset).classification;
}

function isUnitAsset(asset: RawAsset): boolean {
  const cls = classOf(asset);
  return cls === CLS_UNIT || cls === CLS_SQUAD;
}

function isWeaponAsset(asset: RawAsset): boolean {
  const cls = classOf(asset);
  // Match exact "Weapon", or any "<X> Weapon" classification (e.g. "Ranged Weapon").
  return cls === CLS_WEAPON || cls.endsWith(` ${CLS_WEAPON}`);
}

function isEnhancementAsset(asset: RawAsset): boolean {
  return classOf(asset) === CLS_ENHANCEMENT;
}

function isCharacterAsset(asset: RawAsset): boolean {
  const keywords = asObject(asset.keywords);
  if (keywords) {
    for (const list of Object.values(keywords)) {
      for (const kw of asArray(list)) {
        if (typeof kw === "string" && CHAR_CLASSIFICATIONS.has(kw)) return true;
      }
    }
  }
  // Any nested trait classified as Character also flags the unit.
  for (const t of traits(asset)) {
    if (CHAR_CLASSIFICATIONS.has(classOf(t))) return true;
    const dsg = displayName(t);
    if (CHAR_CLASSIFICATIONS.has(dsg)) return true;
  }
  return false;
}

function isWarlordTrait(asset: RawAsset): boolean {
  const { classification, designation } = splitItem(asset);
  if (designation === DSG_WARLORD) return true;
  return classification === CLS_TRAIT && designation === DSG_WARLORD;
}

/** Sum every Weapon/Enhancement-bearing leaf quantity to derive the unit's
 * model count. Falls back to `quantity(unit)` when the tree has no per-model
 * markers (single-model entries). */
function modelCount(unit: RawAsset): number {
  // Rosterizer doesn't carve "model" out separately from "unit" the way
  // BattleScribe does; the unit's own quantity is the model count for
  // squads. For multi-model squads with explicit per-model children, each
  // child unit-class asset's quantity contributes.
  let nested = 0;
  for (const child of included(unit)) {
    if (isUnitAsset(child)) nested += quantity(child);
  }
  return nested > 0 ? nested : quantity(unit);
}

function parseUnit(unit: RawAsset): ParsedUnit {
  const wargear: ParsedWargear[] = [];
  let enhancement_raw_name: string | null = null;
  let enhancement_points: number | null = null;
  let is_warlord = false;

  for (const child of included(unit)) {
    walk(child, (a) => {
      if (isEnhancementAsset(a)) {
        if (enhancement_raw_name === null) {
          enhancement_raw_name = displayName(a);
          enhancement_points = pointsOf(a);
        }
        return;
      }
      if (isWeaponAsset(a)) {
        wargear.push({ raw_name: displayName(a), count: quantity(a) });
      }
    });
  }
  for (const t of traits(unit)) {
    walk(t, (a) => {
      if (isWarlordTrait(a)) is_warlord = true;
    });
  }

  return {
    raw_name: displayName(unit),
    is_character: isCharacterAsset(unit),
    model_count: modelCount(unit),
    points: pointsOf(unit),
    is_warlord,
    enhancement_raw_name,
    enhancement_points,
    wargear,
  };
}

/** Resolve the snapshot Asset tree from an envelope, preferring the explicit
 * `snapshot` field but falling through to the history-present roster. */
function snapshotOf(env: RawEnvelope): RawAsset | null {
  const snap = asObject(env.snapshot);
  if (snap) return snap as RawAsset;
  const history = asObject(env.history) as RawHistory | null;
  const present = history && asObject(history.present);
  if (present) {
    const present_roster = asObject((present as RawHistoryItem).roster);
    if (present_roster) return present_roster as RawAsset;
  }
  return null;
}

function isRosterizerEnvelope(decoded: unknown): decoded is RawEnvelope {
  const env = asObject(decoded) as RawEnvelope | null;
  if (!env) return false;
  if (!asObject(env.rulebook)) return false;
  return snapshotOf(env) !== null;
}

/** Find the first child Asset with the given classification, if any. */
function findChildByClass(asset: RawAsset, cls: string): RawAsset | null {
  for (const c of included(asset)) {
    if (classOf(c) === cls) return c;
  }
  return null;
}

function parseLimit(label: string | null): number | null {
  if (!label) return null;
  const match = POINTS_LIMIT.exec(label);
  if (!match) return null;
  return Number.parseInt(match[1].replace(/,/g, ""), 10);
}

export const rosterizerAdapter: FormatAdapter = {
  id: "rosterizer",

  matches(decoded: unknown): boolean {
    return isRosterizerEnvelope(decoded);
  },

  parse(decoded: unknown): ParsedRoster {
    if (!isRosterizerEnvelope(decoded)) {
      throw new Error("rosterizer: payload is not a Rosterizer roster envelope");
    }
    const snapshot = snapshotOf(decoded);
    if (snapshot === null) {
      throw new Error("rosterizer: envelope has no snapshot or history.present.roster");
    }

    // Treat the snapshot as the roster root regardless of its `item` value —
    // some exports root at `Roster§Roster`, others at the faction itself.
    const root = snapshot;

    // Roster-level metadata children. Faction and detachment come from the
    // first child Asset of their respective classification; battle size the
    // same way. Walk the whole tree (rather than just root.assets.included)
    // so nested-force shapes still pick up the markers.
    let faction_raw_name: string | null = null;
    let detachment_raw_name: string | null = null;
    let battle_size_raw: string | null = null;
    const factions: string[] = [];
    walk(root, (a) => {
      const cls = classOf(a);
      if (cls === CLS_FACTION) {
        const name = displayName(a);
        if (!factions.includes(name)) factions.push(name);
        faction_raw_name ??= name;
      } else if (cls === CLS_DETACHMENT) {
        detachment_raw_name ??= displayName(a);
      } else if (cls === CLS_BATTLE_SIZE) {
        battle_size_raw ??= displayName(a);
      }
    });

    // Allow the rulebook envelope to carry a battle-size override (e.g.
    // `rulebook.notes: "2000 Point limit"`) — strictly optional.
    if (battle_size_raw === null) {
      const rulebook = asObject((decoded as RawEnvelope).rulebook) as RawRulebook | null;
      // Intentionally not reading rulebook.notes — it may carry prose.
      void rulebook;
    }

    // Collect units: any Unit/Squad asset anywhere in the tree, excluding
    // ones nested under another unit (those are attached leaders we'll fold
    // into leader_attachment via the resolver — but ParsedUnit doesn't model
    // them yet, so for v1 every Unit asset becomes a top-level ParsedUnit).
    const units: ParsedUnit[] = [];
    const collectUnits = (a: RawAsset, underUnit: boolean): void => {
      if (isUnitAsset(a) && !underUnit) {
        units.push(parseUnit(a));
        for (const c of included(a)) collectUnits(c, true);
        for (const c of traits(a)) collectUnits(c, true);
        return;
      }
      if (isUnitAsset(a) && underUnit) {
        // Nested unit (leader on a body, body on a leader, etc.) — emit it as
        // its own top-level ParsedUnit so the resolver can match its id and
        // the leader-attachment inference pass can link the two.
        units.push(parseUnit(a));
        return;
      }
      for (const c of included(a)) collectUnits(c, underUnit);
      for (const c of traits(a)) collectUnits(c, underUnit);
    };
    collectUnits(root, false);

    // Roster-level total: prefer an explicit Points stat on the root, else
    // sum every unit's (base + enhancement) contribution.
    const total_reported = pointsOf(root);
    let total_computed = 0;
    for (const u of units) {
      total_computed += u.points ?? 0;
      total_computed += u.enhancement_points ?? 0;
    }

    const env = decoded as RawEnvelope;
    const rulebook = asObject(env.rulebook) as RawRulebook | null;
    const generated_by = rulebook ? asString(rulebook.name) ?? asString(rulebook.url) : null;
    const name = displayName(root) || asString(rulebook?.name) || "Imported roster";

    return {
      name,
      generated_by,
      faction_raw_name,
      detachment_raw_name,
      battle_size_raw,
      declared_limit: parseLimit(battle_size_raw),
      total_reported,
      total_computed,
      units,
      multi_force: factions.length > 1,
    };
  },
};

// Internals re-exported for the symmetric exporter and unit tests.
export const _internals = {
  CLS_ROSTER,
  CLS_FACTION,
  CLS_DETACHMENT,
  CLS_UNIT,
  CLS_WEAPON,
  CLS_ENHANCEMENT,
  CLS_BATTLE_SIZE,
  CLS_TRAIT,
  DSG_WARLORD,
  POINTS_STAT_KEYS,
  splitItem,
  displayName,
  classOf,
  findChildByClass,
};

File: python/src/wh40kdc/imports/rosterizer.py
"""Rosterizer adapter: lower a Rosterizer roster JSON payload to a
``ParsedRoster``.

Rosterizer (https://rosterizer.com) stores a roster as a ``Roster`` envelope
with a recursive ``Asset`` tree under ``snapshot`` (or
``history.present.roster`` as a fallback). Every entity is an ``Asset`` keyed
by ``Classification§Designation`` (e.g. ``"Unit§Tactical Squad"``). Children
sit under ``assets.included`` (game pieces) and ``assets.traits`` (modifiers,
abilities, markers).

**IP safety**: the walk reads an ALLOWLIST — ``item``, ``designation``,
``name``, ``classification``, ``quantity``, ``meta.points``,
``stats.Points.value``, and the recursive ``assets.included``/
``assets.traits`` children. Prose-bearing fields are never touched.

Python mirror of ``tools/src/import/rosterizer.ts``.
"""

from __future__ import annotations

import math
import re
from collections.abc import Callable
from typing import Any

from wh40kdc.imports.adapter import FormatAdapter

# --- 40K rulebook Classification§Designation conventions. --------------------

_CLS_FACTION = "Faction"
_CLS_DETACHMENT = "Detachment"
_CLS_UNIT = "Unit"
_CLS_SQUAD = "Squad"  # alternative unit class some rulebooks use
_CLS_WEAPON = "Weapon"
_CLS_ENHANCEMENT = "Enhancement"
_CLS_BATTLE_SIZE = "Battle Size"
_CLS_TRAIT = "Trait"
_DSG_WARLORD = "Warlord"
_CHAR_CLASSIFICATIONS = frozenset(["Character", "Epic Hero"])

_POINTS_STAT_KEYS = ("Points", "Pts")
_POINTS_LIMIT = re.compile(r"(\d[\d,]*)\s*Point", re.IGNORECASE)

# --- Structural views --------------------------------------------------------


def _as_array(value: Any) -> list[Any]:
    return value if isinstance(value, list) else []


def _as_object(value: Any) -> dict[str, Any] | None:
    return value if isinstance(value, dict) else None


def _as_string(value: Any) -> str | None:
    return value if isinstance(value, str) else None


def _as_number(value: Any) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)) and math.isfinite(value):
        return value
    if isinstance(value, str):
        m = re.match(r"\s*[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?", value)
        if m:
            try:
                n = float(m.group(0))
            except ValueError:
                return None
            return n if math.isfinite(n) else None
    return None


def _split_item(asset: dict[str, Any]) -> tuple[str, str]:
    """Split ``Classification§Designation`` into its two halves. Falls back to
    the raw ``classification``/``designation`` fields when ``item`` is absent."""
    item = _as_string(asset.get("item"))
    if item is not None:
        idx = item.find("§")
        if idx >= 0:
            return item[:idx], item[idx + 1 :]
    return (
        _as_string(asset.get("classification")) or "",
        _as_string(asset.get("designation")) or "",
    )


def _display_name(asset: dict[str, Any]) -> str:
    """A user-facing display name for an asset: ``name`` override beats the
    designation parsed out of the ``item`` key."""
    name = _as_string(asset.get("name"))
    return name if name is not None else _split_item(asset)[1]


def _quantity(asset: dict[str, Any]) -> int:
    n = _as_number(asset.get("quantity"))
    return math.trunc(n) if n is not None and n > 0 else 1


def _included(asset: dict[str, Any]) -> list[dict[str, Any]]:
    a = _as_object(asset.get("assets"))
    return [c for c in _as_array(a.get("included") if a else None) if isinstance(c, dict)]


def _traits(asset: dict[str, Any]) -> list[dict[str, Any]]:
    a = _as_object(asset.get("assets"))
    return [c for c in _as_array(a.get("traits") if a else None) if isinstance(c, dict)]


def _points_of(asset: dict[str, Any]) -> int | None:
    """Points cost from ``stats.Points.value`` (or aliases) / ``meta.points``."""
    stats = _as_object(asset.get("stats"))
    if stats:
        for key in _POINTS_STAT_KEYS:
            stat = _as_object(stats.get(key))
            if stat:
                v = _as_number(stat.get("value"))
                if v is not None:
                    return math.trunc(v)
    meta = _as_object(asset.get("meta"))
    if meta:
        v = _as_number(meta.get("points"))
        if v is not None:
            return math.trunc(v)
    return None


def _walk(asset: dict[str, Any], visit: Callable[[dict[str, Any]], None]) -> None:
    """Depth-first visit of an asset and every included/trait descendant."""
    visit(asset)
    for child in _included(asset):
        _walk(child, visit)
    for child in _traits(asset):
        _walk(child, visit)


def _class_of(asset: dict[str, Any]) -> str:
    return _split_item(asset)[0]


def _is_unit_asset(asset: dict[str, Any]) -> bool:
    return _class_of(asset) in (_CLS_UNIT, _CLS_SQUAD)


def _is_weapon_asset(asset: dict[str, Any]) -> bool:
    cls = _class_of(asset)
    # Match exact "Weapon", or any "<X> Weapon" (e.g. "Ranged Weapon").
    return cls == _CLS_WEAPON or cls.endswith(f" {_CLS_WEAPON}")


def _is_enhancement_asset(asset: dict[str, Any]) -> bool:
    return _class_of(asset) == _CLS_ENHANCEMENT


def _is_character_asset(asset: dict[str, Any]) -> bool:
    keywords = _as_object(asset.get("keywords"))
    if keywords:
        for kw_list in keywords.values():
            for kw in _as_array(kw_list):
                if isinstance(kw, str) and kw in _CHAR_CLASSIFICATIONS:
                    return True
    # Any nested trait classified as Character also flags the unit.
    for t in _traits(asset):
        if _class_of(t) in _CHAR_CLASSIFICATIONS:
            return True
        if _display_name(t) in _CHAR_CLASSIFICATIONS:
            return True
    return False


def _is_warlord_trait(asset: dict[str, Any]) -> bool:
    classification, designation = _split_item(asset)
    if designation == _DSG_WARLORD:
        return True
    return classification == _CLS_TRAIT and designation == _DSG_WARLORD


def _model_count(unit: dict[str, Any]) -> int:
    """Sum nested unit-class asset quantities; fall back to the unit's own
    quantity for single-model entries."""
    nested = 0
    for child in _included(unit):
        if _is_unit_asset(child):
            nested += _quantity(child)
    return nested if nested > 0 else _quantity(unit)


def _parse_unit(unit: dict[str, Any]) -> dict[str, Any]:
    wargear: list[dict[str, Any]] = []
    state: dict[str, Any] = {
        "enhancement_raw_name": None,
        "enhancement_points": None,
        "is_warlord": False,
    }

    def visit(a: dict[str, Any]) -> None:
        if _is_enhancement_asset(a):
            if state["enhancement_raw_name"] is None:
                state["enhancement_raw_name"] = _display_name(a)
                state["enhancement_points"] = _points_of(a)
            return
        if _is_weapon_asset(a):
            wargear.append({"raw_name": _display_name(a), "count": _quantity(a)})

    for child in _included(unit):
        _walk(child, visit)

    def visit_trait(a: dict[str, Any]) -> None:
        if _is_warlord_trait(a):
            state["is_warlord"] = True

    for t in _traits(unit):
        _walk(t, visit_trait)

    return {
        "raw_name": _display_name(unit),
        "is_character": _is_character_asset(unit),
        "model_count": _model_count(unit),
        "points": _points_of(unit),
        "is_warlord": state["is_warlord"],
        "enhancement_raw_name": state["enhancement_raw_name"],
        "enhancement_points": state["enhancement_points"],
        "wargear": wargear,
    }


def _snapshot_of(env: dict[str, Any]) -> dict[str, Any] | None:
    """Resolve the snapshot Asset tree from an envelope, preferring the
    explicit ``snapshot`` field but falling through to the history-present
    roster."""
    snap = _as_object(env.get("snapshot"))
    if snap:
        return snap
    history = _as_object(env.get("history"))
    present = _as_object(history.get("present")) if history else None
    if present:
        present_roster = _as_object(present.get("roster"))
        if present_roster:
            return present_roster
    return None


def _is_rosterizer_envelope(decoded: Any) -> bool:
    env = _as_object(decoded)
    if not env:
        return False
    if not _as_object(env.get("rulebook")):
        return False
    return _snapshot_of(env) is not None


def _parse_limit(label: str | None) -> int | None:
    if not label:
        return None
    match = _POINTS_LIMIT.search(label)
    if not match:
        return None
    return int(match.group(1).replace(",", ""))


def _matches(decoded: Any) -> bool:
    return _is_rosterizer_envelope(decoded)


def _parse(decoded: Any) -> dict[str, Any]:
    if not _is_rosterizer_envelope(decoded):
        raise ValueError("rosterizer: payload is not a Rosterizer roster envelope")
    snapshot = _snapshot_of(decoded)
    if snapshot is None:
        raise ValueError("rosterizer: envelope has no snapshot or history.present.roster")

    # Treat the snapshot as the roster root regardless of its `item` value —
    # some exports root at `Roster§Roster`, others at the faction itself.
    root = snapshot

    # Roster-level metadata children: first child Asset of each classification.
    # Walk the whole tree so nested-force shapes still pick up the markers.
    meta_state: dict[str, Any] = {
        "faction_raw_name": None,
        "detachment_raw_names": [],
        "battle_size_raw": None,
    }
    factions: list[str] = []

    def visit_meta(a: dict[str, Any]) -> None:
        cls = _class_of(a)
        if cls == _CLS_FACTION:
            name = _display_name(a)
            if name not in factions:
                factions.append(name)
            if meta_state["faction_raw_name"] is None:
                meta_state["faction_raw_name"] = name
        elif cls == _CLS_DETACHMENT:
            meta_state["detachment_raw_names"].append(_display_name(a))
        elif cls == _CLS_BATTLE_SIZE:
            if meta_state["battle_size_raw"] is None:
                meta_state["battle_size_raw"] = _display_name(a)

    _walk(root, visit_meta)

    # Collect units: any Unit/Squad asset anywhere in the tree. A unit nested
    # under another unit (leader on a body, etc.) is emitted as its own
    # top-level ParsedUnit so the resolver can match its id and the
    # leader-attachment inference pass can link the two.
    units: list[dict[str, Any]] = []

    def collect_units(a: dict[str, Any], under_unit: bool) -> None:
        if _is_unit_asset(a) and not under_unit:
            units.append(_parse_unit(a))
            for c in _included(a):
                collect_units(c, True)
            for c in _traits(a):
                collect_units(c, True)
            return
        if _is_unit_asset(a) and under_unit:
            units.append(_parse_unit(a))
            return
        for c in _included(a):
            collect_units(c, under_unit)
        for c in _traits(a):
            collect_units(c, under_unit)

    collect_units(root, False)

    # Roster-level total: prefer an explicit Points stat on the root, else
    # sum every unit's (base + enhancement) contribution.
    total_reported = _points_of(root)
    total_computed = 0
    for u in units:
        total_computed += u["points"] or 0
        total_computed += u["enhancement_points"] or 0

    rulebook = _as_object(decoded.get("rulebook"))
    generated_by = None
    if rulebook:
        generated_by = _as_string(rulebook.get("name"))
        if generated_by is None:
            generated_by = _as_string(rulebook.get("url"))
    name = _display_name(root)
    if not name:
        name = (_as_string(rulebook.get("name")) if rulebook else None) or "Imported roster"

    battle_size_raw = meta_state["battle_size_raw"]
    return {
        "name": name,
        "generated_by": generated_by,
        "faction_raw_name": meta_state["faction_raw_name"],
        "detachment_raw_names": meta_state["detachment_raw_names"],
        "battle_size_raw": battle_size_raw,
        "declared_limit": _parse_limit(battle_size_raw),
        "total_reported": total_reported,
        "total_computed": total_computed,
        "units": units,
        "multi_force": len(factions) > 1,
    }


rosterizer_adapter = FormatAdapter(id="rosterizer", matches=_matches, parse=_parse)
